### ISSUE TITLE
Support for PS256 in Java 11

### ISSUE THREAD

**Issue by fschutte** [jwtk/jjwt#542]
According to the documentation starting from Java 11 it should be possible to use signatures with PS256 algorithm, but when I try this it gives an error unless BouncyCastle is used. But I though that starting from Java 11 BouncyCastle would just not be necessary anymore.
Any tips?
`.signWith(signKey, SignatureAlgorithm.PS256)`
gives
`io.jsonwebtoken.security.SignatureException: Unavailable RSA Signature algorithm 'SHA256withRSAandMGF1'. This is not a standard JDK algorithm. Try including BouncyCastle in the runtime classpath.
`
Looks like by using PS256 it tries Signature algorithm SHA256withRSAandMGF1 which seems BouncyCastle specific ?

**Comment by lhazlewood** [comment]
Thanks for the issue!

If the string name we use is BouncyCastle specific, it is unclear to me the alternative name that should be used without BC.  For example, which out of following list should be specified?

https://docs.oracle.com/en/java/javase/11/docs/specs/security/standard-names.html

**Comment by lhazlewood** [comment]
I was thinking that something else in addition to `RSASSA-PSS` was necessary.  I think that may be all that's required, and then specify a `PSSParameterSpec` during signature calculation.  I'll investigate more.

**Comment by lhazlewood** [comment]
Yep, that was all that was required in local testing - to change the SignatureAlgorithm `jcaName` values from their `SHA<n>withRSAandMGF1` names to all just `RSASSA-PSS` (all three exact same jcaName):

https://github.com/jwtk/jjwt/blob/eadf0ce4fc252341f04c50a45aac6f9c670283e7/api/src/main/java/io/jsonwebtoken/SignatureAlgorithm.java#L90-L109

This is because `RsaProvider` already sets up the correct `PSSParameterSpec` instances for each algorithm:

https://github.com/jwtk/jjwt/blob/eadf0ce4fc252341f04c50a45aac6f9c670283e7/impl/src/main/java/io/jsonwebtoken/impl/crypto/RsaProvider.java#L39-L56

Finally, I confirmed that if on JDK 11, the SignatureAlgorithm provider is:

    SunRsaSign version 11

On <= JDK 10, where BouncyCastle is required, with no code changes, the provider changes to:

    BC version 1.6

So BC supports that name as well, and the `SunRsaSign` implementation is always preferred unless unavailable.

We'll leave this ticket open to track the PR for these changes.  Thanks again for the issue!

Implementation note:  It is important to include a test case that uses digest output from JJWT version 0.10.7 using PS256, PS384, and PS512 algorithms.  And then assert that the PR changes with `RSASSA-PSS` can produce and consume these exact same existing digest values to ensure that this change is backwards compatible with existing signed JWTs created with JJWT <= 0.10.7.

**Comment by lhazlewood** [comment]
Second implementation note:

`SignatureAlgorithmTest.bigInteger` is unbearably slow due to the 2nd `certainty` argument of the `BigInteger` constructor.  Since we don't need ideal randomness for these tests (we're just using that method to produce mock values), change the implementation to be the following:

```java
    private static BigInteger bigInteger(int bitLength) {
        BigInteger result = null
        while (result == null || result.bitLength() != bitLength) {
            result = new BigInteger(bitLength, Random.newInstance())
        }
        return result
    }
```

This was much faster and sped up the build significantly.

**Comment by fschutte** [comment]
Hi Les, thanks for picking this up. Do I understand correctly that it could easily be fixed, but it requires a change in the library (namely in SignatureAlgorithm.java)? If so, will it be part of a future new version?

**Comment by lhazlewood** [comment]
@fschutte yep, that's correct!

**Comment by lhazlewood** [comment]
Resolved via #554
